docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+- EMA_RESERVE, only the virtual address range is reserved. No EPC pages will be committed in this mode.  
+- EMA_COMMIT_NOW: reserves and commits physical EPC committed upon allocation. EACCEPT will be done immediately.
+- EMA_COMMIT_ON_DEMAND: EACCEPT is done on demand, see below on committing and uncommitting.
+
+An allocation, once created, will own its address range until the deallocation API, sgx_mm_dealloc, is called upon. No two active allocations can have overlapping address ranges.
+
+**Commit, Uncommit Enclave Memory**
+
+When a page in COMMIT_ON_DEMAND allocations is accessed, a page fault occurs if the page was not yet commited.  The EMM will perform EACCEPT to commit the EPC page on page fault after OS doing EAUG. 
+
+The clients can also call the EMM commit API, sgx_mm_commit, to proactively commit specific sub-regions in a COMMIT_ON_DEMAND allocation to avoid future page fault.
+
+Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page fault, can register a custom handler for page fault at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
+
+Committed pages will stay committed (regardless how they were committed) until the clients calls the uncommit API, sgx_mm_uncommit, on them or the allocation they belong to is deallocated by sgx_mm_dealloc.    
+
+**Modify Page Attributes**
+
+The EMM clients may call sgx_mm_modify/sgx_mm_modify_ex to request permissions and/or page type changes for pages in existing allocations.
+
+
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
+- The EMA list tracks all memory regions in use (reserved, committed, commit-on-demand) in ELRANGE.
+- Ranges in ELRANGE not tracked by an EMA are considered free and ready for new allocations
+- The EMM labels certain EMAs reserved for runtime or its internal usage and make them not accessible from public APIs.
+- A thread calling an EMM API on an EMA with operation pending in another thread will wait until the pending operation is finished. 
+
+**Assumptions:**
+
+- When enclave is loaded, the OS has reserved the whole address range covered by ELRANGE. It is assumed the host app will not remap any part of this reserved range.
+  - In future, when the new feature that allows loading enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
```

Comment:
```suggestion
- The memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC is running low or the cgroups threshold is reached
```

Also regarding "when cgroups threshold reached": how does this apply to Windows?